pkg/dyninst/decode/decoder.go

```diff
@@ -89,15 +82,27 @@ type typeAndAddr struct {
 func (d *Decoder) Decode(
 	event Event,
 	symbolicator symbol.Symbolicator,
-	out io.Writer,
-) (probe ir.ProbeDefinition, err error) {
+) (output []byte, probe ir.ProbeDefinition, err error) {
+	if d.decoderTypes == nil {
+		d.decoderTypes = make(map[ir.TypeID]decoderType, len(d.program.Types))
+		for _, t := range d.program.Types {
+			decoderType, err := newDecoderType(t, d.program.Types)
+			if err != nil {
+				log.Tracef("error getting decoder type for type %s: %w", t.GetName(), err)
+				continue
+			}
+			d.decoderTypes[t.GetID()] = decoderType
+		}
+	}
```

Comment:
I want to push back on making this lazy. Imagine we have a bug where we're going to fail on this. Now every time we get an event we'll hit this code. It will never succeed. But even further, I think it's bad to have this work happening in the data plane path. Right now we construct the decoder before attachment which is a good thing. Also, it feels right to me that we fail with probe diagnostics and in a loud way if we can't construct a decoder.

I think having the opportunity to return an error from decoder construction was a good thing. It is akin to a failure to compile the probe. I see no reason to make the change, and I especially see no reason to couple it with this PR.